modules/renter/proto/downloader.go

```diff
@@ -35,6 +37,17 @@ func (hd *Downloader) Sector(root crypto.Hash) (modules.RenterContract, []byte,
 		return modules.RenterContract{}, nil, errors.New("contract has insufficient funds to support download")
 	}
 
+	// create the download revision
+	rev := newDownloadRevision(hd.contract.LastRevision, sectorPrice)
+
+	// if a SaveFn was provided, call it before doing any further I/O, because
+	// that's when we're most likely to experience failure
+	if hd.SaveFn != nil {
+		if err := hd.SaveFn(rev); err != nil {
+			return modules.RenterContract{}, nil, errors.New("failed to save unsigned revision: " + err.Error())
+		}
+	}
```

Comment:
currently, `Downloader` and `Editor` do not require that a SaveFn is provided. Alternatively, we could force callers to supply it as an argument to `NewEditor`/`NewDownloader`, or we could supply a default `SaveFn` that simply returns `nil`.